File: polyglot/java/profile_parser.java
package polyglot.java;

import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.util.regex.*;

/**
 * Profile Parser for GitHub configuration files.
 * Supports YAML and JSON formats commonly used in .github/config.yaml or similar setups.
 */
public class profile_parser {

    // ==================== DATA MODEL ====================

    public record ProfileConfig(
        String name,
        String bio,
        String location,
        String website,
        List<String> socialLinks,
        Map<String, Object> metadata
    ) {}

    // ==================== PARSER FACTORY ====================

    private static final class ParserFactory {
        public static ProfileConfig parse(String content) throws IOException {
            if (content == null || content.trim().isEmpty()) {
                throw new IllegalArgumentException("Empty input");
            }

            String trimmed = content.trim();
            
            // Detect format by header or structure
            if (trimmed.startsWith("---") && trimmed.contains(":")) {
                return parseYaml(content);
            } else if (trimmed.startsWith("{") || trimmed.startsWith("[{")) {
                return parseJson(content);
            } else {
                throw new IOException("Unknown format: " + trimmed.substring(0, Math.min(50, trimmed.length())));
            }
        }

        private static ProfileConfig parseYaml(String content) throws IOException {
            // Simple YAML parser for common profile fields
            ProfileConfig config = new ProfileConfig(null, null, null, null, null, nullMap());
            
            try (StringReader reader = new StringReader(content)) {
                String line;
                while ((line = reader.readLine()) != null) {
                    line = line.trim();
                    if (line.isEmpty() || line.startsWith("#") || line.startsWith("---")) {
                        continue;
                    }

                    // Parse key: value pairs
                    int colonIndex = line.indexOf(':');
                    if (colonIndex > 0) {
                        String key = line.substring(0, colonIndex).trim();
                        String value = line.substring(colonIndex + 1).trim();

                        switch (key.toLowerCase()) {
                            case "name":
                                config.name = unquote(value);
                                break;
                            case "bio" || case "description":
                                config.bio = unquote(value);
                                break;
                            case "location":
                            case "based_in":
                                config.location = unquote(value);
                                break;
                            case "website" || case "homepage":
                                config.website = unquote(value);
                                break;
                            case "social_links" || case "links":
                                parseSocialLinks(config, value);
                                break;
                            default:
                                // Store unknown keys in metadata
                                if (config.metadata == nullMap()) {
                                    config.metadata = new LinkedHashMap<>();
                                }
                                config.metadata.put(key.toLowerCase(), unquote(value));
                        }
                    } else if (line.startsWith("- ")) {
                        // Handle list items like: - https://twitter.com/username
                        String item = line.substring(2).trim();
                        if (config.socialLinks == nullList()) {
                            config.socialLinks = new ArrayList<>();
                        }
                        config.socialLinks.add(unquote(item));
                    } else if (!line.startsWith("-")) {
                        // Handle nested structures or list items without dash prefix
                        String item = line.trim();
                        if (config.socialLinks == nullList()) {
                            config.socialLinks = new ArrayList<>();
                        }
                        config.socialLinks.add(unquote(item));
                    }
                }
            }

            return config;
        }

        private static ProfileConfig parseJson(String content) throws IOException {
            // Simple JSON parser for profile fields
            ProfileConfig config = new ProfileConfig(null, null, null, null, nullList(), nullMap());
            
            try (StringReader reader = new StringReader(content)) {
                int braceDepth = 0;
                StringBuilder currentKey = new StringBuilder();
                StringBuilder currentValue = new StringBuilder();
                
                for (int i = 0; i < content.length(); i++) {
                    char c = content.charAt(i);

                    if (c == '{') {
                        braceDepth++;
                        continue;
                    } else if (c == '}') {
                        braceDepth--;
                        if (braceDepth == 0) break;
                        continue;
                    }

                    // Handle key-value pairs
                    if (c == '"' && !currentKey.isEmpty()) {
                        String key = currentKey.toString();
                        String val = currentValue.toString().trim();

                        switch (key.toLowerCase()) {
                            case "name":
                                config.name = unquote(val);
                                break;
                            case "bio" || case "description":
                                config.bio = unquote(val);
                                break;
                            case "location":
                            case "based_in":
                                config.location = unquote(val);
                                break;
                            case "website" || case "homepage":
                                config.website = unquote(val);
                                break;
                            case "social_links" || case "links":
                                parseJsonSocialLinks(config, val);
                                break;
                            default:
                                if (config.metadata == nullMap()) {
                                    config.metadata = new LinkedHashMap<>();
                                }
                                config.metadata.put(key.toLowerCase(), unquote(val));
                        }

                        currentKey.setLength(0);
                        currentValue.setLength(0);
                    } else if (c != ' ' && c != '\t' && c != ':' && c != ',' && c != '[' && c != ']') {
                        currentValue.append(c);
                    } else if (c == '"' && !currentKey.isEmpty()) {
                        // End of key or value string
                        String key = currentKey.toString();
                        String val = currentValue.toString().trim();

                        switch (key.toLowerCase()) {
                            case "name":
                                config.name = unquote(val);
                                break;
                            case "bio" || case "description":
                                config.bio = unquote(val);
                                break;
                            case "location":
                            case "based_in":
                                config.location = unquote(val);
                                break;
                            case "website" || case "homepage":
                                config.website = unquote(val);
                                break;
                            case "social_links" || case "links":
                                parseJsonSocialLinks(config, val);
                                break;
                            default:
                                if (config.metadata == nullMap()) {
                                    config.metadata = new LinkedHashMap<>();
                                }
                                config.metadata.put(key.toLowerCase(), unquote(val));
                        }

                        currentKey.setLength(0);
                        currentValue.setLength(0);
                    } else if (c != ' ' && c != '\t' && c != ':' && c != ',' && c != '[' && c != ']') {
                        currentValue.append(c);
                    }
                }
            }

            return config;
        }

        // ==================== HELPER METHODS ====================

        private static void parseSocialLinks(ProfileConfig config, String value) throws IOException {
            if (config.socialLinks == nullList()) {
                config.socialLinks = new ArrayList<>();
            }
            
            // Handle both inline list and string representation
            if (value.startsWith("[")) {
                try {
                    String[] items = value.substring(1, value.length() - 1).split(",");
                    for (String item : items) {
                        config.socialLinks.add(unquote(item.trim()));
                    }
                } catch (Exception e) {
                    // Fallback: treat as single string
                    config.socialLinks.add(unquote(value));
                }
            } else {
                config.socialLinks.add(unquote(value));
            }
        }

        private static void parseJsonSocialLinks(ProfileConfig config, String value) throws IOException {
            if (config.socialLinks == nullList()) {
                config.socialLinks = new ArrayList<>();
            }
            
            // Handle JSON array string like: ["https://twitter.com/x", "https://linkedin.com/in/y"]
            if (value.startsWith("[")) {
                try {
                    String[] items = value.substring(1, value.length() - 1).split(",");
                    for (String item : items) {
                        config.socialLinks.add(unquote(item.trim()));
                    }
                } catch (Exception e) {
                    config.socialLinks.add(unquote(value));
                }
            } else {
                config.socialLinks.add(unquote(value));
            }
        }

        private static String unquote(String value) {
            if (value == null || value.isEmpty()) return "";
            
            // Remove surrounding quotes
            if ((value.startsWith("\"") && value.endsWith("\"")) || 
                (value.startsWith("'") && value.endsWith("'"))) {
                return value.substring(1, value.length() - 1);
            }
            
            return value;
        }

        private static <T> T nullList() {
            return null;
        }

        private static <K, V> Map<K, V> nullMap() {
            return null;
        }

        // ==================== FILE OPERATIONS ====================

        public static ProfileConfig parseFile(String path) throws IOException {
            if (path == null || path.trim().isEmpty()) {
                throw new IllegalArgumentException("Path cannot be empty");
            }

            Path p = Paths.get(path);
            if (!Files.exists(p)) {
                throw new FileNotFoundException("Profile file not found: " + path);
            }

            String content;
            try (BufferedReader reader = Files.newBufferedReader(p, StandardCharsets.UTF_8)) {
                StringBuilder sb = new StringBuilder();
                String line;
                while ((line = reader.readLine()) != null) {
                    sb.append(line).append('\n');
                }
                content = sb.toString();
            }

            return parse(content);
        }

        public static ProfileConfig parseFile(Path path) throws IOException {
            if (path == null || !Files.exists(path)) {
                throw new FileNotFoundException("Profile file not found: " + path);
            }

            String content;
            try (BufferedReader reader = Files.newBufferedReader(path, StandardCharsets.UTF_8)) {
                StringBuilder sb = new StringBuilder();
                String line;
                while ((line = reader.readLine()) != null) {
                    sb.append(line).append('\n');
                }
                content = sb.toString();
            }

            return parse(content);
        }

        // ==================== VALIDATION ====================

        public static boolean validate(ProfileConfig config, String purpose) {
            if (config == null) {
                System.err.println("Validation failed: Config is null");
                return false;
            }

            List<String> errors = new ArrayList<>();

            // Check required fields based on purpose
            switch (purpose.toLowerCase()) {
                case "public":
                case "default":
                    if (config.name == null || config.name.trim().isEmpty()) {
                        errors.add("Name is required for public profiles");
                    }
                    break;
                
                case "minimal":
                case "internal":
                    // Minimal validation
                    break;

                default:
                    // Default to minimal
                    break;
            }

            if (!errors.isEmpty()) {
                System.err.println("Validation errors for '" + purpose + "' profile:");
                for (String error : errors) {
                    System.err.println("  - " + error);
                }
                return false;
            }

            System.out.println("Validation passed for '" + purpose + "' profile.");
            return true;
        }

        // ==================== OUTPUT ====================

        public static String format(ProfileConfig config) {
            if (config == null) return "null";

            StringBuilder sb = new StringBuilder();
            sb.append("Profile Configuration:\n");
            sb.append("=====================\n\n");

            appendIfPresent(sb, "Name", config.name);
            appendIfPresent(sb, "Bio/Description", config.bio);
            appendIfPresent(sb, "Location", config.location);
            appendIfPresent(sb, "Website", config.website);

            if (config.socialLinks != nullList()) {
                sb.append("\nSocial Links:\n");
                for (String link : config.socialLinks) {
                    sb.append("  - ").append(link).append('\n');
                }
            }

            if (config.metadata != nullMap() && !config.metadata.isEmpty()) {
                sb.append("\nMetadata:\n");
                for (var entry : config.metadata.entrySet()) {
                    sb.append("  ").append(entry.getKey()).append(": ")
                      .append(entry.getValue()).append('\n');
                }
            }

            return sb.toString();
        }

        private static void appendIfPresent(StringBuilder sb, String label, Object value) {
            if (value != null && !Objects.equals(value, "")) {
                sb.append(label).append(": ").append(value).append('\n');
            }
        }

        // ==================== DEMO / ENTRY POINT ====================

        public static void main(String[] args) throws Exception {
            System.out.println("GitHub Profile Parser - Demo\n");
            System.out.println("=".repeat(40));

            // Sample YAML config (typical .github/config.yaml format)
            String sampleYaml = """
            ---
            name: "Jane Doe"
            bio: "Senior Java Engineer | Open Source Contributor"
            location: "San Francisco, CA"
            website: "https://janedoe.dev"
            
            social_links:
              - "https://twitter.com/janedoe"
              - "https://github.com/janedoe"
              - "https://linkedin.com/in/janedoe"
              
            metadata:
              email: "jane@example.com"
              timezone: "America/Los_Angeles"
            """;

            // Sample JSON config
            String sampleJson = """
            {
                "name": "John Smith",
                "bio": "Full-Stack Developer",
                "location": "New York, NY",
                "website": "https://johnsmith.io",
                "social_links": [
                    "https://twitter.com/johnsmith",
                    "https://github.com/johnsmith"
                ],
                "metadata": {
                    "email": "john@example.com",
                    "timezone": "America/New_York"
                }
            }
            """;

            // Demo 1: Parse YAML string
            System.out.println("\n[Demo 1] Parsing YAML String...");
            try {
                ProfileConfig yamlProfile = parseYaml(sampleYaml);
                System.out.println(format(yamlProfile));
            } catch (Exception e) {
                System.err.println("Error: " + e.getMessage());
            }

            // Demo 2: Parse JSON string
            System.out.println("\n[Demo 2] Parsing JSON String...");
            try {
                ProfileConfig jsonProfile = parseJson(sampleJson);
                System.out.println(format(jsonProfile));
            } catch (Exception e) {
                System.err.println("Error: " + e.getMessage());
            }

            // Demo 3: Parse from file (create temp file for demo)
            System.out.println("\n[Demo 3] Parsing from File...");
            try {
                Path tempFile = Files.createTempFile("profile-", ".yaml");
                Files.writeString(tempFile, sampleYaml);
                
                ProfileConfig fileProfile = parseFile(tempFile.toString());
                System.out.println(format(fileProfile));
                
                // Cleanup
                Files.deleteIfExists(tempFile);
            } catch (Exception e) {
                System.err.println("Error: " + e.getMessage());
            }

            // Demo 4: Validation
            System.out.println("\n[Demo 4] Validating Profile...");
            try {
                ProfileConfig testProfile = parseYaml(sampleYaml);
                validate(testProfile, "public");
            } catch (Exception e) {
                System.err.println("Error: " + e.getMessage());
            }

            // Demo 5: Edge cases
            System.out.println("\n[Demo 5] Edge Cases...");
            
            // Empty input
            try {
                ProfileConfig empty = parse("");
                System.out.println("Empty string result: " + format(empty));
            } catch (Exception e) {
                System.err.println("Empty string error: " + e.getMessage());
            }

            // Whitespace only
            try {
                ProfileConfig whitespace = parse("   \n\n  ");
                System.out.println("Whitespace result: " + format(whitespace));
            } catch (Exception e) {
                System.err.println("Whitespace error: " + e.getMessage());
            }

            // Malformed JSON
            try {
                ProfileConfig malformed = parseJson("{invalid json}");
                System.out.println("Malformed result: " + format(malformed));
            } catch (Exception e) {
                System.err.println("Malformed error: " + e.getMessage());
            }

            // Demo 6: Real-world usage pattern
            System.out.println("\n[Demo 6] Real-World Usage Pattern...");
            try {
                // Simulate reading from a config directory
                String[] samplePaths = {"/tmp/profile.yaml", "/home/user/.github/config.json"};
                
                for (String path : samplePaths) {
                    System.out.println("\nChecking: " + path);
                    if (Files.exists(Paths.get(path))) {
                        ProfileConfig profile = parseFile(path);
                        boolean valid = validate(profile, "public");
                        if (valid) {
                            System.out.println("  Status: Ready for deployment");
                        } else {
                            System.out.println("  Status: Needs fixes before deployment");
                        }

File: polyglot/rust/profile_parser.rs
use serde::{Deserialize, Serialize};
use std::fs;
use std::path::PathBuf;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GitHubProfile {
    pub name: String,
    #[serde(default)]
    pub username: Option<String>,
    #[serde(default)]
    pub email: Option<String>,
    #[serde(default)]
    pub bio: Option<String>,
    #[serde(default)]
    pub website: Option<String>,
    #[serde(default)]
    pub location: Option<String>,
    #[serde(default)]
    pub avatar_url: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProfileConfig {
    pub github: GitHubProfile,
    #[serde(default)]
    pub metadata: Metadata,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct Metadata {
    #[serde(default)]
    pub version: String,
    #[serde(default)]
    pub created_at: Option<String>,
    #[serde(default)]
    pub updated_at: Option<String>,
}

impl GitHubProfile {
    pub fn validate(&self) -> Vec<ValidationError> {
        let mut errors = Vec::new();

        if self.name.is_empty() {
            errors.push(ValidationError::Field("name".into(), "must not be empty"));
        }

        if let Some(ref username) = self.username {
            if username.chars().count() > 50 {
                errors.push(ValidationError::Field("username".into(), "max 50 characters"));
            }
        }

        if let Some(ref email) = self.email {
            if !email.contains('@') || !email.contains('.') {
                errors.push(ValidationError::Field("email".into(), "must look like a valid email"));
            }
        }

        errors
    }

    pub fn to_json(&self, pretty: bool) -> Result<String, serde_json::Error> {
        if pretty {
            serde_json::to_string_pretty(self)
        } else {
            serde_json::to_string(self)
        }
    }

    pub fn from_json(data: &str) -> Result<Self, serde_json::Error> {
        serde_json::from_str(data)
    }
}

#[derive(Debug)]
pub enum ValidationError {
    Field(String, String),
    Config(String),
}

impl std::fmt::Display for ValidationError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ValidationError::Field(field, msg) => write!(f, "field '{}' error: {}", field, msg),
            ValidationError::Config(msg) => write!(f, "config error: {}", msg),
        }
    }
}

impl std::error::Error for ValidationError {}

pub struct ProfileParser {
    config_path: Option<PathBuf>,
    auto_detect_format: bool,
}

impl Default for ProfileParser {
    fn default() -> Self {
        Self {
            config_path: None,
            auto_detect_format: true,
        }
    }
}

impl ProfileParser {
    pub fn new(config_path: Option<PathBuf>) -> Self {
        Self {
            config_path,
            auto_detect_format: true,
        }
    }

    pub fn with_auto_detect(mut self, detect: bool) -> Self {
        self.auto_detect_format = detect;
        self
    }

    pub fn parse(&self, source: &str) -> Result<ProfileConfig, ParseError> {
        let config: ProfileConfig = match serde_yaml::from_str(source) {
            Ok(c) => c,
            Err(_) => {
                return Err(ParseError::Format("YAML".into(), "failed to parse YAML"));
            }
        };

        if !config.github.name.is_empty() || config.github.username.is_some() {
            Ok(config)
        } else {
            Err(ParseError::EmptyProfile)
        }
    }

    pub fn load(&self, path: &PathBuf) -> Result<ProfileConfig, ParseError> {
        let source = fs::read_to_string(path).map_err(|e| {
            ParseError::Io(format!("failed to read file '{}': {}", path.display(), e))
        })?;

        self.parse(&source)
    }

    pub fn save(&self, config: &ProfileConfig, path: &PathBuf) -> Result<(), SaveError> {
        let json = config.github.to_json(true)?;
        fs::write(path, json).map_err(|e| {
            SaveError::Io(format!("failed to write '{}': {}", path.display(), e))
        })?;

        Ok(())
    }
}

#[derive(Debug)]
pub enum ParseError {
    Format(String, String),
    EmptyProfile,
    Io(String),
}

impl std::fmt::Display for ParseError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ParseError::Format(format_name, msg) => write!(f, "format '{}' error: {}", format_name, msg),
            ParseError::EmptyProfile => write!(f, "profile is empty or missing required fields"),
            ParseError::Io(msg) => write!(f, "{}", msg),
        }
    }
}

impl std::error::Error for ParseError {}

#[derive(Debug)]
pub enum SaveError {
    Io(String),
}

impl std::fmt::Display for SaveError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            SaveError::Io(msg) => write!(f, "{}", msg),
        }
    }
}

impl std::error::Error for SaveError {}

fn detect_format(source: &str) -> Option<&'static str> {
    if source.starts_with("{") || source.contains(": ") {
        Some("JSON")
    } else if source.starts_with("-" | source.lines().any(|l| l.trim_start().starts_with("-"))) {
        Some("YAML")
    } else {
        None
    }
}

pub fn detect_format_from_source(source: &str) -> Option<&'static str> {
    if let Some(format) = detect_format(source) {
        return Some(format);
    }

    // Fallback to YAML as default for GitHub configs
    Some("YAML")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_valid_profile() {
        let profile = ProfileConfig {
            github: GitHubProfile {
                name: "Jane Doe".into(),
                username: Some("janedoe".into()),
                email: Some("jane@example.com".into()),
                bio: Some("Rust enthusiast".into()),
                website: Some("https://example.com".into()),
                location: Some("San Francisco, CA".into()),
                avatar_url: None,
            },
            metadata: Metadata {
                version: "1.0.0".into(),
                created_at: Some("2024-01-01T00:00:00Z".into()),
                updated_at: Some("2024-01-01T00:00:00Z".into()),
            },
        };

        let errors = profile.github.validate();
        assert!(errors.is_empty());
    }

    #[test]
    fn test_empty_name_error() {
        let profile = ProfileConfig {
            github: GitHubProfile {
                name: String::new(),
                ..Default::default()
            },
            metadata: Metadata::default(),
        };

        let errors = profile.github.validate();
        assert!(errors.iter().any(|e| e.to_string().contains("name")));
    }

    #[test]
    fn test_json_roundtrip() {
        let profile = GitHubProfile {
            name: "Test User".into(),
            ..Default::default()
        };

        let json = profile.to_json(true).unwrap();
        let parsed: GitHubProfile = serde_json::from_str(&json).unwrap();
        
        assert_eq!(parsed.name, "Test User");
    }

    #[test]
    fn test_parse_error_display() {
        let err = ParseError::Format("YAML".into(), "invalid syntax".into());
        assert!(err.to_string().contains("format"));
    }
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("cognis-digital: GitHub Profile Parser");
    println!("=====================================");

    // Demo: Create a sample profile
    let sample_profile = ProfileConfig {
        github: GitHubProfile {
            name: "Example Developer".into(),
            username: Some("example-dev".into()),
            email: Some("dev@example.com".into()),
            bio: Some("Full-stack developer passionate about Rust and web technologies.".into()),
            website: Some("https://github.com/example-dev".into()),
            location: Some("Remote, Worldwide".into()),
            avatar_url: None,
        },
        metadata: Metadata {
            version: "1.0.0".into(),
            created_at: Some(format!(
                "{}",
                chrono::Utc::now().format("%Y-%m-%dT%H:%M:%SZ")
            )),
            updated_at: None,
        },
    };

    println!("\n--- Sample Profile ---");
    println!("{}", sample_profile.github.to_json(true)?);

    // Demo: Validate the profile
    let errors = sample_profile.github.validate();
    if errors.is_empty() {
        println!("\n✓ Validation passed!");
    } else {
        for error in &errors {
            println!("✗ {}", error);
        }
    }

    // Demo: Parse from JSON string
    let json_input = r#"{
        "github": {
            "name": "Another User",
            "username": "another-user"
        },
        "metadata": {}
    }"#;

    println!("\n--- Parsing JSON Input ---");
    match sample_profile.github.from_json(json_input) {
        Ok(parsed) => println!("✓ Parsed successfully: {}", parsed.name),
        Err(e) => println!("✗ Parse error: {}", e),
    }

    // Demo: Detect format automatically
    let yaml_input = r#"- github:
    name: YAML User
    username: yaml-user
- metadata: {}"#;

    println!("\n--- Format Detection ---");
    if let Some(detected) = detect_format_from_source(yaml_input) {
        println!("✓ Detected format: {}", detected);
    } else {
        println!("? Could not auto-detect format, defaulting to YAML");
    }

    Ok(())
}